pkg/resource/deploy/step.go

```diff
@@ -709,6 +709,135 @@ func (s *RefreshStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
+}
+
+func NewImportStep(plan *Plan, reg RegisterResourceEvent, new *resource.State) Step {
+	contract.Assert(new != nil)
+	contract.Assert(new.URN != "")
+	contract.Assert(new.ID != "")
+	contract.Assert(new.Custom)
+	contract.Assert(!new.Delete)
+	contract.Assert(!new.External)
+
+	return &ImportStep{
+		plan: plan,
+		reg:  reg,
+		new:  new,
+	}
+}
+
+func NewImportReplacementStep(plan *Plan, reg RegisterResourceEvent, original, new *resource.State) Step {
+	contract.Assert(original != nil)
+	contract.Assert(new != nil)
+	contract.Assert(new.URN != "")
+	contract.Assert(new.ID != "")
+	contract.Assert(new.Custom)
+	contract.Assert(!new.Delete)
+	contract.Assert(!new.External)
+
+	return &ImportStep{
+		plan:      plan,
+		reg:       reg,
+		original:  original,
+		new:       new,
+		replacing: true,
+	}
+}
+
+func (s *ImportStep) Op() StepOp {
+	if s.replacing {
+		return OpImportReplacement
+	}
+	return OpImport
+}
+
+func (s *ImportStep) Plan() *Plan                                  { return s.plan }
+func (s *ImportStep) Type() tokens.Type                            { return s.new.Type }
+func (s *ImportStep) Provider() string                             { return s.new.Provider }
+func (s *ImportStep) URN() resource.URN                            { return s.new.URN }
+func (s *ImportStep) Old() *resource.State                         { return s.old }
+func (s *ImportStep) New() *resource.State                         { return s.new }
+func (s *ImportStep) Res() *resource.State                         { return s.new }
+func (s *ImportStep) Logical() bool                                { return !s.replacing }
+func (s *ImportStep) Diffs() []resource.PropertyKey                { return s.diffs }
+func (s *ImportStep) DetailedDiff() map[string]plugin.PropertyDiff { return s.detailedDiff }
+
+func (s *ImportStep) Apply(preview bool) (resource.Status, StepCompleteFunc, error) {
+	complete := func() { s.reg.Done(&RegisterResult{State: s.new}) }
+
+	// Read the current state of the resource to import. If the provider does not hand us back any inputs for the
+	// resource, it probably needs to be updated. If the resource does not exist at all, fail the import.
+	prov, err := getProvider(s)
+	if err != nil {
+		return resource.StatusOK, nil, err
+	}
+	read, rst, err := prov.Read(s.new.URN, s.new.ID, nil, nil)
+	if err != nil {
+		if initErr, isInitErr := err.(*plugin.InitError); isInitErr {
+			s.new.InitErrors = initErr.Reasons
+		} else {
+			return rst, nil, err
+		}
+	}
+	if read.Outputs == nil {
+		return rst, nil, errors.Errorf("resource '%v' does not exist", s.new.ID)
+	}
+	if read.Inputs == nil {
+		return resource.StatusOK, nil, errors.Errorf(
+			"provider does not support importing resources; please try updating the '%v' plugin",
+			s.new.URN.Type().Package())
+	}
+	s.new.Outputs = read.Outputs
+
+	// Magic up an old state so the frontend can display a proper diff.
```

Comment:
Might be nice to explain the magic you are doing here, so it was clear why we choose the values we did.  I think the state you are creating here is the logical state that would be returned by a Read step, and I think that's the correct thing to do, but I am not sure.

Reply:
Yes, that's right. I will expand the comment.